### ISSUE TITLE
Design Qwen-Mathia v2 general on Qwen3.5-4B-Base

### ISSUE THREAD

**Issue by murillo128** [murillo128/mathia#55]
Research umbrella: #2
Conceptual-specialist line: #29
Precedent/training baseline: #47 / PR #48
General corpus baseline: #44
Agnostic OpenAlex supplement: merged through #42 / PR #51
Riemann specialization: **out of scope**

# Goal

Design the next **domain-agnostic Qwen-Mathia checkpoint**, explicitly building on the successful Qwen-Mathia v1 training contract while adapting it to the much larger general Mathia corpus and to `Qwen/Qwen3.5-9B-Base`.

This is a **design + CPU materialization issue**. It should end with an exact, auditable v2-general training contract and frozen general-training manifest, ready for a separate GPU execution issue. It does **not** run the full SFT or publish weights.

The intended model lineage is:

```text
Qwen/Qwen3.5-9B-Base
        |
        |  expanded general Mathia SFT
        v
qwen-mathia-v2-general
        |
        `--> later A/B: base vs Mathia-v2 through qwen-lean + DeepSeek + Lean
```

Do **not** include Riemann-Mathia objects here. A later Riemann specialization should be a separate model/experiment (for example `qwen-mathia-riemann-v1`) and must preserve the general-vs-domain-specialization distinction.

## Explicit sequencing decision

#47 already recorded a user-authorized override of the older #29/#32 sequencing guardrail. The user has now made the next sequencing decision explicitly on **2026-08-23**:

1. train a stronger **general Mathia v2** first;
2. wait for the current qwen-lean SFT to finish;
3. then run the discriminating A/B using the new v2 model, not the small v1 checkpoint, against both qwen-lean and DeepSeek with Lean verification.

Reason: a negative result from v1 would be ambiguous because its training corpus was tiny, while a positive result would still lead to v2 training. Therefore v1 is retained as a historical/engineering baseline, not the current scientific gate.

This sequencing choice does **not** turn training loss, fluent prose, or teacher similarity into evidence of Mathia capability. The later solver A/B remains the scientific test.

# V1 evidence to inherit

#47 / PR #48 established a clean engineering baseline:

- exact pretrained base/tokenizer revision pinned;
- frozen corpus and deterministic training manifest before GPU work;
- source objects used as mathematical context/provenance, **not standalone optimizer targets**;
- accepted `interpretation` and `synthesis` objects as optimizer targets;
- completion-only causal-LM loss: prompt/context masked, response + EOS supervised;
- NF4 QLoRA / PEFT rather than full-weight training;
- no packing;
- no silent truncation;
- deterministic ordering/seed and exact token/loss-mask evidence;
- operational smoke before full GPU execution;
- intermediate checkpoints retained without subjective checkpoint selection;
- terminal artifact published as a PEFT adapter, not a redundant merged full model;
- immutable Hugging Face revision and hash verification after publication;
- conservative licensing metadata because corpus source licenses remain mixed;
- no scientific theorem-proving/reasoning claim from training alone.

V1 outcome for reference:

```text
base: Qwen/Qwen3-8B-Base@49e3418fbbbca6ecbdf9608b4d22e5a407081db4
targets: 98 interpretations + 18 syntheses
max sequence: 727 tokens (bound 768)
training: 4 epochs / 60 optimizer steps
example presentations: 464
supervised response/EOS token presentations: 52,868
adapter: murillo2000/qwen3-8b-base-mathia-v1
HF revision: a0dade8a2ce5222b1222f762434ca89528587f69
```

These are engineering priors, not numbers to copy blindly.

# General corpus G-v2

Construct one frozen **general-only** training workload from the union of:

## A. #44 frozen general corpus

```text
release: agnostic-mathia-full-v1
freeze: freeze_eeeeb89af3d2ac75d1ff5dad5623b63d1d24dfbddb965beca2f1c4aac9f9867f
reviewed-content freeze: review_content_d1d1d7152fa2c2ddd3a4f6d26a4fa4b3f6d64129392b7c79ea72f125b5d95c0b
contract: mathia-interchange-v1
```

#47 found 98 accepted/eligible interpretation targets + 18 accepted synthesis targets in this release. Recompute from the frozen manifest; do not trust the historical count blindly.

## B. frozen agnostic OpenAlex supplement from #42

```text
release: agnostic-mathia-openalex-supplement-v1
path: experiments/agnostic_mathia_corpus/openalex_supplement_v1/
freeze: agnostic_openalex_supplement_a1aa591df034db64d5ce0271df0da570e3aaf470ac49e5cc4014b66181bf0e33
```

The published report records 300 semantic units / derivatives, with 295 accepted, 3 quarantined and 2 rejected; 595 canonical objects are eligible overall. It also states explicitly that Riemann records/artifacts are not members of this release lineage.

## Corpus construction rules

Before training design is accepted:

1. verify both frozen parents and all relevant file hashes;
2. preserve original object IDs, source lineage and quality decisions;
3. deduplicate across #44 and the OpenAlex supplement by exact identity/content hashes and source/version lineage;
4. add a near-duplicate diagnostic for rendered optimizer targets, but do not automatically delete mathematically distinct examples merely because prose is similar;
5. exclude all `rejected`, `quarantined`, `evaluation_only`, Riemann-domain, and non-training-eligible objects;
6. preserve the v1 rule that `source` objects are context/provenance rather than standalone optimizer targets;
7. supervise accepted/eligible `interpretation` and `synthesis` responses only;
8. never let the 595 eligible supplement-object count be mistaken for 595 supervised targets;
9. record exact target counts by parent release, role and source after dedupe;
10. freeze a compact `G-v2` manifest before GPU execution.

Expected order of magnitude before cross-release dedupe is roughly 116 #44 conceptual targets plus 295 accepted supplement interpretations. This is only a design sanity check; the exact manifest is authoritative.

# Frozen model ancestor: change from v1

Primary parent model:

```text
Qwen/Qwen3.5-9B-Base
```

Use the **Base** checkpoint, not `Qwen3.5-9B` post-trained/instruct and not Qwen-Mathia-v1 as a weight parent. Qwen-Mathia-v1 remains a historical control; v2 starts cleanly from the new common base so the later A/B is interpretable.

Before design close:

- pin exact immutable model + tokenizer revision;
- preserve upstream license/model-card evidence;
- verify current `transformers`/PEFT/bitsandbytes compatibility;
- inspect exact module names and parameter ownership in the language path;
- record total/trainable adapter parameter counts for the proposed LoRA targets.

Qwen3.5-9B-Base is not architecturally identical to Qwen3-8B: it has a hybrid language stack and a vision encoder. Therefore **do not copy v1's `q_proj/k_proj/v_proj/...` target list blindly**.

The v2 design must target only the intended **language-model** projections and keep the vision path frozen/untrained. If the architecture needs DeltaNet-specific projection targets, enumerate and test them explicitly. Do not use a broad `all-linear` selector if it would silently capture vision modules.

# Prompt/rendering contract: change from v1

Keep `mathia-interchange-v1` as the semantic corpus contract, but adapt the model wrapper to the new parent model.

Qwen3.5-9B-Base's upstream model card explicitly supports LoRA-style PEFT with its official control/chat tokens. The **primary v2 design should therefore use the official Qwen3.5 chat template deterministically** unless CPU/preflight evidence finds a concrete incompatibility.

Conceptually:

```text
user message:
  mathematical material + Mathia task

assistant message:
  accepted Mathia interpretation/synthesis
```

Loss remains:

```text
system/user/template/prompt tokens -> masked
assistant Mathia response          -> supervised
assistant EOS/end token            -> supervised
```

Requirements:

- no semantic rewriting of the frozen corpus;
- no prompt augmentation/paraphrase multiplication;
- no assistant-prefix leakage into the supervised target;
- exact rendered bytes + label boundaries reproducible;
- later Base-vs-Mathia A/B must use the **same public prompt/template lane** for both checkpoints.

Do not retain v1's no-chat-template rule merely for historical symmetry when the new upstream base was trained with dedicated control tokens for PEFT.

# Sequence-length audit: keep the invariant, change the number

V1's `768` limit was a measured consequence of its tiny workload, not a project constant.

For G-v2:

1. tokenize every exact rendered target with the pinned Qwen3.5 tokenizer;
2. report prompt, supervised-response and total token distributions separately by role and parent corpus;
3. record p50/p90/p95/p99/max and the concrete longest object IDs;
4. choose the smallest practical bound containing **every** selected target;
5. `truncation = false` remains mandatory;
6. if even one accepted target cannot fit the intended 20 GB QLoRA lane at a technically reasonable bound, surface a design/runtime blocker rather than silently drop or shorten it.

Packing remains `false` in the primary lane unless this design produces a strong, auditable reason to change it. With a few hundred heterogeneous mathematical examples, semantic isolation and simple masking are more valuable than squeezing maximal throughput.

# Training recipe: mostly inherit v1, but do not inherit four epochs

Use v1 as the engineering prior:

```text
base load:             4-bit NF4
compute dtype:         bfloat16
double quantization:   true
PEFT:                   LoRA
LoRA rank:              16
LoRA alpha:             32
LoRA dropout:           0.0
bias:                   none
gradient checkpointing:true
optimizer:              paged_adamw_8bit
learning_rate prior:    5e-5
weight_decay:           0.0
max_grad_norm:          1.0
lr schedule:            cosine
micro batch:            1
grad accumulation:      8
packing:                 false
truncation:              false
seed:                    0
```

The **exact language-only target modules** are a v2 design output, not inherited text.

## Epoch/exposure change

Do **not** reuse `epochs = 4` by default.

V1 needed four passes over only 116 optimizer examples and finished at 60 optimizer steps. G-v2 is expected to contain roughly 3.5× as many optimizer examples before dedupe, and its responses may also be materially longer. Four epochs would therefore multiply supervised-token exposure far beyond v1 without evidence.

Primary design prior:

```text
terminal training exposure: 1 full epoch over frozen G-v2
```

With an effective batch of 8 and ~400 examples this is already on the order of v1's optimizer-step count. The exact token audit must report:

- unique prompt tokens;
- unique supervised tokens;
- total tokens per one corpus pass;
- projected optimizer steps and token presentations for 0.25 / 0.5 / 1.0 epoch.

Freeze the terminal exposure **before GPU training**. Do not tune epochs from downstream qwen-lean/DeepSeek outcomes or subjective generations. If the CPU audit gives a concrete reason that one pass is inappropriate, the design may choose another exposure, but it must document the quantitative reason and precommit it.

Prefer recoverable checkpoints around 25%, 50%, and 100% of the terminal exposure. The root/published adapter is the precommitted terminal checkpoint. Intermediate checkpoints are evidence, not an invitation to pick whichever later scores best without a new preregistered comparison.

# GPU target and preflight

Primary hardware target remains one **NVIDIA RTX 4000 Ada Generation 20 GB** because v1 established the project workflow there.

However, Qwen3.5-9B-Base includes a larger/hybrid text stack and a vision encoder. The execution issue must perform a fresh QLoRA memory/numerics smoke; v1's 4.46 GB headroom is not transferable evidence.

Preflight must prove at least:

- exact base/tokenizer revision;
- text-only forward/backward path;
- vision parameters frozen and absent from adapter targets;
- finite loss and non-zero LoRA gradients;
- only intended adapter parameters change;
- completion-only loss mask correctness;
- no sequence truncation;
- save/reload identity;
- peak memory with a safe margin at the frozen max sequence length.

If one 20 GB Ada cannot support the frozen data/model contract, stop and record the concrete resource blocker. Prefer moving to more memory over shrinking/dropping training examples or silently changing the parent model.

# Publication contract: inherit v1

If the later execution succeeds, publish primarily as a PEFT adapter under `murillo2000`, not merged full weights.

Provisional model name:

```text
murillo2000/qwen3.5-9b-base-mathia-v2
```

Freeze the exact final repository name before execution/publication.

Carry forward the v1 requirements:

- adapter safetensors/config;
- exact training config and G-v2 manifest/freeze IDs;
- checkpoint/hash inventory;
- base/tokenizer immutable revisions;
- package/CUDA/GPU/runtime evidence;
- role/token/exposure counts;
- response-only loss semantics;
- no Riemann data claim;
- no theorem-proving/reasoning claim from training alone;
- conservative source-license handling (`license: other` unless a fresh audit justifies something more specific);
- clean-cache download from the immutable Hub revision and exact hash/reload verification.

Do not upload raw/restricted source material, base-model copies, secrets or caches.

# Downstream scientific handoff — not executed here

The reason for v2 is the later A/B once the current qwen-lean SFT is complete.

Primary comparison:

```text
Qwen3.5-9B-Base  -> generate matched intuitions
qwen-mathia-v2   -> generate matched intuitions
                         |
              +----------+----------+
              |                     |
          qwen-lean               DeepSeek
              |                     |
              +-------- Lean --------+
```

The important evidence is whether Mathia-v2 intuition improves **both** formal workers under matched budgets. Transfer to DeepSeek is especially useful for distinguishing general mathematical utility from solver-specific prompt optimization.

Qwen-Mathia-v1 may be retained as an optional historical control if cheap, but it must not delay the primary Base-vs-v2 comparison.

Do not run this A/B in the design or training execution issue. Keep training and scientific validation separable.

# Design deliverables

This issue should produce, without full GPU SFT:

1. a compact v2-general design document or experiment-local design artifact;
2. exact immutable Qwen3.5-9B-Base model/tokenizer revisions;
3. exact G-v2 parent freeze bindings;
4. deterministic cross-parent dedupe report;
5. frozen G-v2 optimizer manifest with object IDs/roles/hashes/lineage;
6. exact Qwen3.5 rendered prompt/response boundaries;
7. full CPU token-length and supervised-token audit;
8. exact language-only LoRA target-module list and adapter parameter count;
9. frozen max sequence length;
10. frozen optimizer/training-exposure recipe;
11. expected steps/token presentations/checkpoint steps;
12. publication naming/provenance contract;
13. bounded GPU preflight checklist for the later execution issue;
14. explicit handoff to the later Base-vs-Mathia-v2 × {qwen-lean, DeepSeek} A/B.

Keep implementation local to this experiment. Do not create a general training platform/DSL merely because v2 has more data.

# Independent design review

Before closing, run a fresh read-only review checking at minimum:

- no Riemann records entered G-v2;
- exact #44 + agnostic-supplement freeze lineage;
- no rejected/quarantined/evaluation leakage;
- source-context vs optimizer-target semantics preserved;
- dedupe decisions are auditable and do not collapse genuinely distinct mathematics;
- official Qwen3.5 template + response-only mask are correct;
- vision path is excluded from optimization;
- LoRA targets match the actual Qwen3.5 architecture rather than v1 names copied by rote;
- no truncation;
- epoch/token exposure is justified from the new workload rather than historical habit;
- publication and downstream A/B claims remain properly scoped.

# Exit decision

Close with exactly one:

- `QWEN_MATHIA_V2_GENERAL_DESIGN_READY` — open a separate GPU train/publish execution issue using the frozen design;
- `REVISE_QWEN_MATHIA_V2_GENERAL_DESIGN` — concrete design defects remain;
- `QWEN35_COMPATIBILITY_OR_CAPACITY_BLOCKER` — the selected base/runtime cannot satisfy the required training contract without a material redesign.

No full SFT, Hugging Face publication, qwen-lean/DeepSeek inference, Lean proof experiment, Riemann specialization, RL, or weight merging is authorized by this design issue.

**Comment by murillo128** [comment]
## Authoritative backbone amendment — 2026-08-25

The Qwen-Mathia v2 backbone is now **frozen to `Qwen/Qwen3.5-4B-Base`**, superseding every remaining `Qwen3.5-9B-Base` reference in the original issue body.

Use the same exact immutable base/tokenizer revision already frozen by `murillo128/qwen-lean#78`:

```text
model_id:           Qwen/Qwen3.5-4B-Base
model_revision:     1001bb4d826a52d1f399e183466143f4da7b741b
tokenizer_id:       Qwen/Qwen3.5-4B-Base
tokenizer_revision: 1001bb4d826a52d1f399e183466143f4da7b741b
```

This is a deliberate project-level deployment decision, not a temporary capacity fallback. The intended shared-foundation architecture is:

```text
Qwen3.5-4B-Base  (one common resident base)
        |
        +--> Mathia LoRA
        +--> qwen-lean-planner LoRA
        `--> qwen-lean LoRA
```

Rationale:

- qwen-lean #78 already selected this 4B foundation and reports that it outperformed Qwen3.5-9B-Base on both the historical miniF2F and frozen Riemann casting comparisons;
- qwen-lean-planner #80 is already defined as a separate descendant of the same 4B Base;
- keeping Mathia on that same exact ancestor makes a single-base / multi-LoRA inference deployment possible on one RTX 4000 Ada 20 GB, avoiding repeated base-model load/unload between Mathia -> planner -> prover passes;
- the scientific comparison remains Base-4B vs Mathia-4B under matched prompting and downstream formal-worker budgets. This decision does not by itself claim that 4B is generally more capable than 9B.

### Consequent design changes

Treat the following as authoritative replacements for the older 9B text:

- parent model: `Qwen/Qwen3.5-4B-Base@1001bb4d826a52d1f399e183466143f4da7b741b`;
- provisional publication repo: `murillo2000/qwen3.5-4b-base-mathia-v2`;
- downstream base control: exact same Qwen3.5-4B-Base revision;
- architecture inspection / LoRA targeting: Qwen3.5-4B text decoder only, with vision frozen/excluded;
- deployment preflight must additionally test that Mathia's adapter can coexist with the planner and qwen-lean adapters over a **single resident base** under the intended inference runtime, with measured VRAM/KV-cache headroom on the 20 GB Ada;
- prefer LoRA target coverage compatible with Qwen3.5's complete fused text projection groups (`q/k/v`, DeltaNet input groups, output projection, and MLP projections) after exact module-tree verification; do not sacrifice Mathia training semantics merely for runtime convenience.

All corpus, response-only loss, no-truncation, exposure, audit, publication-integrity, Riemann-exclusion, and later scientific-validation requirements from #55 remain unchanged unless they specifically referred to the superseded 9B ancestor.

The issue title has also been updated to `Design Qwen-Mathia v2 general on Qwen3.5-4B-Base`.
